pkg/util/grpc.go

```diff
@@ -2,23 +2,45 @@
 
 package util
 
+import (
+	"errors"
+	"time"
+)
+
+var ErrCloseAndExhaustTimedOut = errors.New("timed out waiting to exhaust stream after calling CloseSend, will continue exhausting stream in background")
+
 type Stream[T any] interface {
 	CloseSend() error
 	Recv() (T, error)
 }
 
+// CloseAndExhaust closes and then tries to exhaust stream. This ensures:
+//   - the gRPC library can release any resources associated with the stream (see https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream)
+//   - instrumentation middleware correctly observes the end of the stream, rather than reporting it as "context canceled"
+//
+// Note that this method may block for up to three seconds if the stream has not already been exhausted.
+// If the stream has not been exhausted after this time, it will return ErrCloseAndExhaustTimedOut and continue exhausting the stream in the background.
 func CloseAndExhaust[T any](stream Stream[T]) error {
-	err := stream.CloseSend()
+	err := stream.CloseSend() //nolint:forbidigo // This is the one place we want to call this method.
 	if err != nil {
 		return err
 	}
 
-	// Exhaust the stream to ensure:
-	// - the gRPC library can release any resources associated with the stream (see https://pkg.go.dev/google.golang.org/grpc#ClientConn.NewStream)
-	// - instrumentation middleware correctly observes the end of the stream, rather than reporting it as "context canceled"
-	for err == nil {
-		_, err = stream.Recv()
-	}
+	done := make(chan struct{})
 
-	return nil
+	go func() {
+		for {
+			if _, err := stream.Recv(); err != nil {
```

Comment:
After 3 seconds the function will return but this goroutine will continue to consume in the background. What if the consume blocks forever? We're going to leak goroutines. So question is: do we ensure that the stream context is canceled on the client side (this side) so that the `Recv()` will unblock and goroutine terminate?

Reply:
For almost all cases where we now use `CloseAndExhaust`, the streaming call is made in the context of an incoming HTTP or gRPC request (eg. querier evaluating a query), and that context will be cancelled when the response is returned. If for some reason that context isn't already being cancelled today, then we're potentially leaking the stream, but then `CloseAndExhaust` should help mitigate that by exhausting the stream and then terminating.

The only exception to this case is `frontendSchedulerWorker` which is used by query-frontends to queue requests to query-schedulers. In this case, the context will always be [cancelled](https://github.com/grafana/mimir/blob/df782798746bcab6361d34e14e2b4ffd33644ca6/pkg/frontend/v2/frontend_scheduler_worker.go#L277) after `CloseAndExhaust` returns.